File: c_dut/src/runner.c
#include "runner.h"
#include "model.h"
#include <stdio.h>
#include <string.h>

void run_config_make_run_id(const RunConfig *config, char *buf, size_t buflen) {
    char bk_str[32];
    bound_k_to_string(config->bound_k, bk_str, sizeof(bk_str));
    
    snprintf(buf, buflen, "%s_%s_%s_%llu_%s",
             config->seed_id,
             policy_to_string(config->policy),
             bk_str,
             (unsigned long long)config->schedule_seed,
             fault_mode_to_string(config->fault_mode));
}

int execute_run(const Seed *seed, const RunConfig *config, 
                const char *out_log_path, RunResult *out_result) {
    NvmeLiteModel model;
    Scheduler scheduler;
    Logger logger;
    
    model_init(&model);
    scheduler_init(&scheduler, config->policy, config->bound_k, config->schedule_seed);
    logger_init(&logger);
    
    size_t n_cmds = seed->n_commands;
    char run_id[512];
    run_config_make_run_id(config, run_id, sizeof(run_id));
    
    size_t submit_window = submit_window_value(config->submit_window);
    
    /* Write header */
    logger_write_header(&logger,
                        run_id,
                        config->seed_id,
                        config->schedule_seed,
                        config->policy,
                        config->bound_k,
                        config->fault_mode,
                        n_cmds,
                        config->submit_window,
                        config->scheduler_version,
                        config->git_commit);
    
    /* Tracking */
    size_t next_cmd = 0;
    uint32_t pending_peak = 0;
    
    /* Fault injection tracking */
    size_t step_count = 0;
    size_t fault_step = (config->fault_mode != FAULT_NONE) ? n_cmds / 2 : (size_t)-1;
    int fault_injected = 0;
    int stop_submits = 0;  /* Set to 1 after TIMEOUT injection */
    
    /* BATCHED policy tracking */
    int batch_remaining = 0;
    const int batch_size = 4;
    
    /* Interleaved submit/complete loop */
    uint32_t pending_buf[MAX_PENDING];
    
    while (1) {
        size_t pending_count = model_pending_count(&model);

        #if INJECT_BUG_ID == 1
        int submit_ok = (pending_count <= submit_window) && (next_cmd < n_cmds) && !stop_submits;
        #else
        int submit_ok = (pending_count < submit_window) && (next_cmd < n_cmds) && !stop_submits;
        #endif

        int complete_ok = (pending_count > 0);
        
        if (!submit_ok && !complete_ok) {
            break;
        }
        
        /* Decide: submit or complete? */
        int do_complete;
        
        /* For BATCHED: if we're in a burst, force complete */
        if (config->policy == POLICY_BATCHED && batch_remaining > 0) {
            do_complete = 1;
        } else if (submit_ok && complete_ok) {
            /* Use RNG bit to decide */
            uint64_t bit = scheduler_next_bit(&scheduler);
            do_complete = (bit == 1);
        } else if (complete_ok) {
            do_complete = 1;
        } else {
            do_complete = 0;
        }
        
        if (do_complete) {
            /* Check fault injection first */
            if (!fault_injected && step_count >= fault_step) {
                if (config->fault_mode == FAULT_TIMEOUT) {
                    /* Get pending and timeout the first one */
                    size_t n_pending = model_get_pending_canonical(&model, pending_buf, MAX_PENDING);
                    if (n_pending > 0) {
                        uint32_t timeout_cmd_id = pending_buf[0];
                        Status timeout_status = STATUS_TIMEOUT;
                        CommandResult result;
                        if (model_complete(&model, timeout_cmd_id, &timeout_status, &result)) {
                            logger_log_complete(&logger, result.cmd_id, result.status, result.output);
                        }
                    }
                    fault_injected = 1;
                    stop_submits = 1;  /* No more SUBMITs after TIMEOUT */
                    step_count++;
                    continue;
                }
                else if (config->fault_mode == FAULT_RESET) {
                    uint32_t pending_before = model_reset(&model);
                    logger_log_reset(&logger, "INJECTED", pending_before);
                    fault_injected = 1;
                    break;
                }
            }
            
            /* Normal completion */
            size_t n_pending = model_get_pending_canonical(&model, pending_buf, MAX_PENDING);
            if (n_pending > 0) {
                /* BATCHED: start new burst if not in one */
                if (config->policy == POLICY_BATCHED && batch_remaining == 0) {
                    batch_remaining = (int)n_pending < batch_size ? (int)n_pending : batch_size;
                }
                
                Decision decision;
                if (scheduler_pick_next(&scheduler, pending_buf, n_pending, &decision)) {
                    CommandResult result;
                    if (model_complete(&model, decision.cmd_id, NULL, &result)) {
                        logger_log_complete(&logger, result.cmd_id, result.status, result.output);
                        /* Decrement batch counter for BATCHED policy */
                        if (config->policy == POLICY_BATCHED && batch_remaining > 0) {
                            batch_remaining--;
                        }
                    }
                }
            }
            step_count++;
        } else {
            /* Submit next command */
            const Command *cmd = &seed->commands[next_cmd];
            uint32_t cmd_id;
            int is_fence;
            uint32_t fence_id;
            
            model_submit(&model, cmd, &cmd_id, &is_fence, &fence_id);
            logger_log_submit(&logger, cmd_id, command_type_name(cmd->type));
            
            if (is_fence) {
                logger_log_fence(&logger, fence_id);
            }
            
            next_cmd++;
            
            uint32_t current = (uint32_t)model_pending_count(&model);
            if (current > pending_peak) {
                pending_peak = current;
            }
        }
    }
    
    /* Write run end */
    uint32_t pending_left = (uint32_t)model_pending_count(&model);
    uint32_t final_peak = pending_peak > model_pending_peak(&model) ? 
                          pending_peak : model_pending_peak(&model);
    
    logger_log_run_end(&logger, pending_left, final_peak);
    
    /* Write log to file */
    if (logger_write_to_file(&logger, out_log_path) != 0) {
        fprintf(stderr, "Error: Cannot write log to %s\n", out_log_path);
        logger_free(&logger);
        return -1;
    }
    
    /* Fill result */
    snprintf(out_result->run_id, sizeof(out_result->run_id), "%s", run_id);
    out_result->pending_left = pending_left;
    out_result->pending_peak = final_peak;
    out_result->had_reset = model_had_reset(&model);
    out_result->commands_lost = model_commands_lost(&model);
    
    logger_free(&logger);
    return 0;
}

File: src/runner.rs
//! Runner: Executes a single run with given parameters
//!
//! Coordinates model, scheduler, and logger to produce deterministic runs.

use crate::logging::{FaultMode, Logger, SerializedSchedule, SubmitWindow};
use crate::model::{NvmeLiteModel, Status};
use crate::scheduler::{BoundK, Policy, Scheduler};
use crate::seed::Seed;
use anyhow::Result;
use std::path::Path;

/// Configuration for a single run
#[derive(Debug, Clone)]
pub struct RunConfig {
    pub seed_id: String,
    pub schedule_seed: u64,
    pub policy: Policy,
    pub bound_k: BoundK,
    pub fault_mode: FaultMode,
    pub submit_window: SubmitWindow,
    pub scheduler_version: String,
    pub git_commit: String,
    pub dump_schedule: bool,
}

impl RunConfig {
    /// Generate deterministic run_id from parameters
    pub fn run_id(&self) -> String {
        format!(
            "{}_{}_{}_{}_{}",
            self.seed_id, self.policy, self.bound_k, self.schedule_seed, self.fault_mode
        )
    }
}

/// Result of a run (for quick access without parsing log)
#[derive(Debug)]
pub struct RunResult {
    pub run_id: String,
    pub pending_left: u32,
    pub pending_peak: u32,
    pub had_reset: bool,
    pub commands_lost: u32,
}

/// Execute a single run
pub fn execute_run(
    seed: &Seed,
    config: &RunConfig,
    out_log: &Path,
    out_schedule: Option<&Path>,
) -> Result<RunResult> {
    let mut model = NvmeLiteModel::new();
    let mut scheduler = Scheduler::new(config.policy, config.bound_k, config.schedule_seed);
    let mut logger = Logger::new();
    let mut schedule = SerializedSchedule::new(
        &seed.seed_id,
        config.schedule_seed,
        config.policy,
        config.bound_k,
        config.fault_mode,
    );

    let n_cmds = seed.commands.len();
    let run_id = config.run_id();
    let submit_window = config.submit_window.value();

    // Write header (with submit_window)
    logger.write_header_with_window(
        &run_id,
        &seed.seed_id,
        config.schedule_seed,
        config.policy,
        config.bound_k,
        config.fault_mode,
        n_cmds,
        &config.scheduler_version,
        &config.git_commit,
        config.submit_window,
    );

    // Tracking
    let mut next_cmd: usize = 0;
    let mut pending_peak: u32 = 0;

    // Helper to submit one command
    let do_submit = |model: &mut NvmeLiteModel, logger: &mut Logger, idx: usize| {
        let command = &seed.commands[idx];
        let (cmd_id, is_fence, fence_id) = model.submit(command.clone());
        logger.log_submit(cmd_id, command.type_name());
        if is_fence {
            if let Some(fid) = fence_id {
                logger.log_fence(fid);
            }
        }
    };

    // No initial fill - start with interleaved loop immediately
    // This allows RNG to affect pending_peak from the start

    // Interleaved submit/complete loop
    let mut step_count = 0;
    let fault_step = if config.fault_mode != FaultMode::NONE {
        Some(n_cmds / 2)
    } else {
        None
    };
    #[allow(unused_assignments)]
    let mut _fault_injected = false;
    let mut stop_submits = false; // Set to true after TIMEOUT injection

    // BATCHED policy tracking
    let mut batch_remaining: usize = 0;
    const BATCH_SIZE: usize = 4;

    loop {
        let pending_count = model.pending_count();
        let submit_ok = pending_count < submit_window && next_cmd < n_cmds && !stop_submits;
        let complete_ok = pending_count > 0;

        if !submit_ok && !complete_ok {
            break;
        }

        // Decide: submit or complete?
        let do_complete = if config.policy == Policy::BATCHED && batch_remaining > 0 {
            // During a burst, force COMPLETE
            true
        } else if submit_ok && complete_ok {
            // Use RNG bit to decide
            let bit = scheduler.next_bit();
            bit == 1
        } else if complete_ok {
            true
        } else {
            false
        };

        if do_complete {
            // Check fault injection first
            if let Some(fs) = fault_step {
                if step_count >= fs && !_fault_injected {
                    match config.fault_mode {
                        FaultMode::TIMEOUT => {
                            let pending = model.get_pending_canonical();
                            if let Some(&cmd_id) = pending.first() {
                                if let Some(result) = model.complete(cmd_id, Some(Status::TIMEOUT))
                                {
                                    logger.log_complete(
                                        result.cmd_id,
                                        result.status,
                                        result.output,
                                    );
                                    schedule.add_fault("TIMEOUT", step_count);
                                }
                            }
                            _fault_injected = true;
                            stop_submits = true; // No more SUBMITs after TIMEOUT
                            step_count += 1;
                            continue;
                        }
                        FaultMode::RESET => {
                            let pending_before = model.reset();
                            logger.log_reset("INJECTED", pending_before);
                            schedule.add_fault("RESET", step_count);
                            _fault_injected = true;
                            break;
                        }
                        FaultMode::NONE => {}
                    }
                }
            }

            // Normal completion
            let pending = model.get_pending_canonical();

            // BATCHED: start new burst if not in one
            if config.policy == Policy::BATCHED && batch_remaining == 0 && !pending.is_empty() {
                batch_remaining = std::cmp::min(BATCH_SIZE, pending.len());
            }

            if let Some(decision) = scheduler.pick_next(&pending) {
                if let Some(result) = model.complete(decision.cmd_id, None) {
                    logger.log_complete(result.cmd_id, result.status, result.output);
                    schedule.add_complete(decision.pick_index);

                    // Decrement batch counter for BATCHED policy
                    if config.policy == Policy::BATCHED && batch_remaining > 0 {
                        batch_remaining -= 1;
                    }
                }
            }
            step_count += 1;
        } else {
            // Submit next command
            do_submit(&mut model, &mut logger, next_cmd);
            next_cmd += 1;
            let current = model.pending_count() as u32;
            if current > pending_peak {
                pending_peak = current;
            }
        }
    }

    // Write run end - use our tracked peak (more accurate with interleaving)
    let pending_left = model.pending_count() as u32;
    // Take max of model's peak and our tracked peak
    let final_peak = std::cmp::max(pending_peak, model.pending_peak());
    logger.log_run_end(pending_left, final_peak);

    // Write outputs
    logger.write_to_file(out_log)?;

    if config.dump_schedule {
        if let Some(schedule_path) = out_schedule {
            schedule.write_to_file(schedule_path)?;
        }
    }

    Ok(RunResult {
        run_id,
        pending_left,
        pending_peak: final_peak,
        had_reset: model.had_reset(),
        commands_lost: model.commands_lost(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::seed::Command;

    fn test_seed() -> Seed {
        Seed {
            seed_id: "test".to_string(),
            commands: vec![
                Command::WRITE {
                    lba: 0,
                    len: 4,
                    pattern: 42,
                },
                Command::READ { lba: 0, len: 4 },
                Command::FENCE,
                Command::WRITE {
                    lba: 8,
                    len: 2,
                    pattern: 7,
                },
            ],
        }
    }

    #[test]
    fn test_fifo_run() {
        let seed = test_seed();
        let config = RunConfig {
            seed_id: "test".to_string(),
            schedule_seed: 0,
            policy: Policy::FIFO,
            bound_k: BoundK::Infinite,
            fault_mode: FaultMode::NONE,
            submit_window: SubmitWindow::Infinite,
            scheduler_version: "test".to_string(),
            git_commit: "none".to_string(),
            dump_schedule: false,
        };

        let tmp = std::env::temp_dir().join("test_fifo.log");
        let result = execute_run(&seed, &config, &tmp, None).unwrap();

        assert_eq!(result.pending_left, 0);
        assert!(result.pending_peak >= 1);

        let log = std::fs::read_to_string(&tmp).unwrap();
        assert!(log.contains("RUN_HEADER("));
        assert!(log.contains("RUN_END(pending_left=0"));
    }

    #[test]
    fn test_reset_run() {
        let seed = test_seed();
        let config = RunConfig {
            seed_id: "test".to_string(),
            schedule_seed: 0,
            policy: Policy::FIFO,
            bound_k: BoundK::Infinite,
            fault_mode: FaultMode::RESET,
            submit_window: SubmitWindow::Infinite,
            scheduler_version: "test".to_string(),
            git_commit: "none".to_string(),
            dump_schedule: false,
        };

        let tmp = std::env::temp_dir().join("test_reset.log");
        let result = execute_run(&seed, &config, &tmp, None).unwrap();

        assert!(result.had_reset);

        let log = std::fs::read_to_string(&tmp).unwrap();
        assert!(log.contains("RESET(reason=INJECTED"));
    }

    #[test]
    fn test_submit_window() {
        let seed = test_seed();
        let config = RunConfig {
            seed_id: "test".to_string(),
            schedule_seed: 42,
            policy: Policy::RANDOM,
            bound_k: BoundK::Infinite,
            fault_mode: FaultMode::NONE,
            submit_window: SubmitWindow::Finite(2),
            scheduler_version: "test".to_string(),
            git_commit: "none".to_string(),
            dump_schedule: false,
        };

        let tmp = std::env::temp_dir().join("test_window.log");
        let result = execute_run(&seed, &config, &tmp, None).unwrap();

        assert_eq!(result.pending_left, 0);
        // With window=2, peak should be <= 2
        assert!(result.pending_peak <= 2);
    }
}
